### PR TITLE
download: Add fallback to private GitHub repo

### PR DESCRIPTION
**Pull request by juliaogris**
If download of GitHub release fails with HTTP status 404 and the
`HERMIT_GITHUB_TOKEN` environment variable isn't empty attempt to treat
the URL like that of a private GitHub repo:

* Authenticate HTTP requests with Auth header
* Get `assetID` via Github API
* Download via `assetID`

I'm not sure if you want me to write a CI test for this - I'd probably 
have to setup probably a private repo with a release under the `cashapp` org?

### CODE REVIEW

File: cache/cache.go [hunk 1]
```diff
@@ -289,3 +303,92 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
...
+	if g.asset, err = url.PathUnescape(m[4]); err != nil {
+		return nil, false
+	}
+	return g, true
+}
+
+// GHRelease is a minimal type for GitHub releases meta information
+// retrieved via the GitHub API see:
+// https://docs.github.com/en/rest/reference/repos#list-releases
+type GHRelease struct {
```
Comment (alecthomas) [inline cache/cache.go:343]: Make this private?

File: cache/cache.go [hunk 2]
```diff
@@ -289,3 +303,92 @@ func (p2 *progressWriterAt) WriteAt(p []byte, off int64) (n int, err error) {
...
+	TagName string    `json:"tag_name"`
+	Assets  []GHAsset `json:"assets"`
+}
+
+// GHAsset is a minimal type for assets in the GitHub releases
+// meta information retrieved via the GitHub API see:
+// https://docs.github.com/en/rest/reference/repos#list-releases
+type GHAsset struct {
```
Comment (alecthomas) [inline cache/cache.go:351]: And this?

**Review by alecthomas — APPROVED** [review]

File: cache/cache.go (hunk 1 above)
Comment (juliaogris) [reply cache/cache.go]: oh, I thought this had to be public for JSON unmarshalling. TIL.

**Comment by alecthomas** [comment]
✋

**Pull request merged**